src/structarray.jl

```diff
@@ -453,6 +453,22 @@ function StructArrayStyle{S, M}(::Val{N}) where {S, M, N}
     return StructArrayStyle{T, N}()
 end
 
+_dimmax(a::Integer, b::Integer) = max(a, b)
+_dimmax(::Type{Any}, ::Integer) = Any
+_dimmax(::Integer ,::Type{Any}) = Any
+_dimmax(::Type{Any} ,::Type{Any}) = Any
+
+# StructArrayStyle is a wrapped style.
+# Here we try our best to resolve style conflict.
+function BroadcastStyle(b::AbstractArrayStyle{M}, a::StructArrayStyle{S, N}) where {S, N, M}
+    S′ = Broadcast.result_style(S(), b)
+    if S′ isa StructArrayStyle # avoid double wrap
+        return typeof(S′)(Val{_dimmax(N, M)}())
+    end
+    StructArrayStyle{typeof(S′), _dimmax(N, M)}()
+end
+BroadcastStyle(::StructArrayStyle, ::DefaultArrayStyle) = Unknown()
```

Comment:
It's probably my lack of knowledge on the `Base.Broadcast` module, but why do we need to special-case `DefaultArrayStyle`? This may be naïve on my side, but if the fallback is general, shouldn't it also cover the default case?

Reply:
`DefaultArrayStyle` is special as any `AbstractArrayStyle` beats it by default.
But some user-defined style, e.g. `StaticArrayStyle`, might be beaten by it.
Without this the following will cause a style-confict error.
```julia
julia> S1 = StructArrays.StructArrayStyle{StaticArrays.StaticArrayStyle{1},1}()
StructArrays.StructArrayStyle{StaticArrays.StaticArrayStyle{1}, 1}()

julia> S2 = Broadcast.DefaultArrayStyle{1}()
Base.Broadcast.DefaultArrayStyle{1}()

julia> Broadcast.result_style(S1,S2)
ERROR: conflicting broadcast rules defined
  Broadcast.BroadcastStyle(::StructArrays.StructArrayStyle{StaticArrays.StaticArrayStyle{1}, 1}, ::Base.Broadcast.DefaultArrayStyle{1}) = StructArrays.StructArrayStyle{StaticArrays.StaticArrayStyle{1}, 1}()
  Broadcast.BroadcastStyle(::Base.Broadcast.DefaultArrayStyle{1}, ::StructArrays.StructArrayStyle{StaticArrays.StaticArrayStyle{1}, 1}) = StructArrays.StructArrayStyle{Base.Broadcast.DefaultArrayStyle{1}, 1}()
One of these should be undefined (and thus return Broadcast.Unknown).
```